R/spaces.R

```diff
@@ -0,0 +1,152 @@
+#' }
+#'
+#' @param name (character) Space name.
+#' @param spaces_key (character) String containing a spaces access key. If
+#'   missing, defaults to value stored in an environment variable \code{DO_SPACES_ACCESS_KEY}.
+#' @param spaces_secret (character) String containing the secret associated
+#'   with the spaces key. If missing, defaults to value stored in an environment
+#'   variable \code{DO_SPACES_SECRET_KEY}.
+#' @param ... Additional arguments passed down to \code{\link[aws.s3]{bucketlist}},
+#'   \code{\link[aws.s3]{get_bucket}}, \code{\link[aws.s3]{put_bucket}} functions
+#'   from the \code{aws.s3} package.
+#' @examples \dontrun{
+#' # list spaces
+#' spaces()
+#'
+#' # obtain spaces as a list of space objects
+#' res <- spaces()
+#'
+#' # print space summary using a space object
+#' summary(res[['my_space_name']])
+#'
+#' # create a new space
+#' space_create('new_space_name')
+#' }
+
+#' @importFrom aws.s3 bucketlist
+#' @keywords internal
+spaces_GET <- function(spaces_key = NULL, spaces_secret = NULL, ...) {
+
+  spaces_key <- check_space_access(spaces_key)
+  spaces_secret <- check_space_secret(spaces_secret)
+
+  res <- aws.s3::s3HTTP(verb = "GET",
+                        region = NULL,
+                        key = spaces_key,
+                        secret = spaces_secret,
+                        base_url = spaces_base,
+                        ...)
+
+  return(res)
+}
+
+#' @export
+#' @rdname spaces
+spaces <- function(spaces_key = NULL, spaces_secret = NULL, ...) {
+  res <- spaces_GET(spaces_key = spaces_key, spaces_secret = spaces_secret, ...)
+  spaces <- lapply(res$Buckets, structure, class = "space")
+  setNames(spaces, vapply(res$Buckets, function(x) x$Name, character(1)))
+}
+
+#' @importFrom aws.s3 get_bucket
+#' @keywords internal
+space_info <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
+  if (is.null(name)) stop("Please specify the space name")
+  spaces_key <- check_space_access(spaces_key)
+  spaces_secret <- check_space_secret(spaces_secret)
+
+  space_info <- get_bucket(name,
+                           region = NULL,
+                           check_region = FALSE,
+                           key = spaces_key,
+                           secret = spaces_secret,
+                           base_url = spaces_base,
+                           max = Inf,
+                           ...)
+
+  return(space_info)
+}
+
+space_size <- function(space_info) {
+  # grab the sizes from each file (unit is bytes)
+  sizes <- vapply(space_info, function(x) x$Size, numeric(1))
+
+  # compute total size (convert to gb)
+  sum(sizes) * 1e-09
+}
+
+space_files <- function(space_info) {
+  # remove entries with size 0 (those are nested directories)
+  length(lapply(space_info, function(x) x[x$Size > 0]))
+}
+
+#' Create a new space
+#' @importFrom aws.s3 put_bucket
+#' @export
+#' @rdname spaces
+space_create <- function(name, spaces_key = NULL, spaces_secret = NULL, ...) {
```

Comment:
don't know if that makes sense for spaces, def. feel free to not change it

Reply:
I'm inclined to not give `name` a default because like Amazon S3, a space name must be unique across all spaces defined in digital ocean, not just the spaces within a user's account.  While the actual chance of users getting the same randomly generated name are quite small, I'm still weary of generating a random name as default.

Reply:
ok